File: Part5/C#/Statements.cs
using System;

class Program
{
    static void Main()
    {
        // Most programs allow this operation, but where is the result? We can't use it!
        // The C# compiler prevents this.
        5 + 3;

        // Simple assignment of a constant and a variable to another variable:
        int a = 1;
        int b = a;
        Console.WriteLine($"Our two variables {a} and {b}");

        // Assignment errors - This code will likely go in its own separate program.
        int c;
        a = c;
        c = c + 1;
        // The C# compiler prevents this.
        1 = a;
        Console.WriteLine($"Our incorrectly assigned variables {a} and {c}");

        // Now let's look at assigning the result of basic operations
        // Assign the result of a mathematical operation
        int value1 = 5;
        int value2 = 6;
        int sumValue = value1 + value2;
        Console.WriteLine($"First value ( {value1} ) plus second value ( {value2} ) is {sumValue}");

        // Assign the result of a comparison operation
        int value3 = 7;
        int value4 = 8;
        bool comparisonValue = (value3 >= value4);
        Console.WriteLine($"First value ( {value3} ) greater than or equal to second value ( {value4} ) is {comparisonValue}");

        // Assign the result of a logical operation
        bool value5 = true;
        bool value6 = false;
        bool logicalValue = value5 || value6;
        Console.WriteLine($"First value ( {value5} ) ORed with second value ( {value6} ) is {logicalValue}");

        // Storing a basic complex operation, multiple additions
        sumValue = 1 + 2 + 3;
        Console.WriteLine($"The sum of 1 + 2 + 3 is {sumValue}");

        // Complex logical operation with differing operation priorities
        bool loyaltyMember = false;
        bool purchased5Coffees = true;
        bool haveCoupon = true;
        bool couponNotExpired = true;
        bool freeCoffee = loyaltyMember && purchased5Coffees || haveCoupon && couponNotExpired;
        Console.WriteLine($"Customer gets a free coffee {freeCoffee}");

        // The code above is basically the same as the following, except the extra variables
        // Sometimes breaking apart complex statements can help to self document
        bool loyaltyAchieved = loyaltyMember && purchased5Coffees;
        bool haveValidCoupon = haveCoupon && couponNotExpired;
        freeCoffee = loyaltyAchieved && haveValidCoupon;
        Console.WriteLine($"Customer gets a free coffee {freeCoffee}");

        // This code shows how a value will be computed incorrectly 
        // with default operator precedence
        double itemPrice = 99.99;
        double itemShipping = 9.99;
        int purchaseQuantity = 5;
        double totalCost = itemPrice + itemShipping * purchaseQuantity;
        Console.WriteLine($"Total item cost is {totalCost:F2}");

        // We can correct this with parenthesis which force operations to complete first
        totalCost = (itemPrice + itemShipping) * purchaseQuantity;
        Console.WriteLine($"Total item cost is {totalCost:F2}");

        // Parenthesis can be used to clarify precedence without 
        // having to know what the actual operator precedence is.
        freeCoffee = ((loyaltyMember == true) && (purchased5Coffees == true)) || ((haveCoupon == true) && (couponNotExpired == true));

        // Some common errors, I can't really write these as a structured 
        // language program, so these are more pseudo code examples.
        // Do you want to assign 5 to a and b, or do you want to check if b is equal to 5?
        // C# does not allow a boolean result assigned to an int, so this is less confusing.
        a = b = 5;
        // vs
        bool aBool;
        aBool = (b == 5);

        // Spot the incorrect operator.
        freeCoffee = ((loyaltyMember = true) && (purchased5Coffees == true)) || ((haveCoupon == true) && (couponNotExpired == true));

        // This will not be isWeekend is true if day is either Saturday or Sunday
        int SATURDAY = 6;
        int SUNDAY = 7;
        int day = 6;
        // C# compiler prevents this code.
        bool isWeekend = (day == SATURDAY || SUNDAY);

        // This is what you should do
        isWeekend = (day == SATURDAY) || (day == SUNDAY);
    }
}

File: Part5/Rust/Statements.rs
fn main() {
    // Most programs allow this operation, but where is the result? We can't use it!
    5 + 3;

    // Simple assignment of a constant and a variable to another variable:
    let a = 1;
    let b = a;
    println!( "Our two variables {} and {}", a, b );

    // Assignment errors - This code will likely go in its own separate program.
    // Rust's compiler will not allow this whole block.
    let c;
    let a = c;
    c = c + 1;
    1 = a;
    println!( "Our incorrectly assigned variables {} and {}", a, c );

    // Now let's look at assigning the result of basic operations
    // Assign the result of a mathematical operation
    let value1 = 5;
    let value2 = 6;
    let sum_value = value1 + value2;
    println!( "First value ( {} ) plus second value ( {} ) is {}", value1, value2, sum_value );

    // Assign the result of a comparison operation
    let value3 = 7;
    let value4 = 8;
    let comparison_value = value3 >= value4;
    println!( "First value ( {} ) greater than or equal to second value ( {} ) is {}", value3, value4, comparison_value );

    // Assign the result of a logical operation
    let value5 = true;
    let value6 = false;
    let logical_value = value5 || value6;
    println!( "First value ( {} ) ORed with second value ( {} ) is {}", value5, value6, logical_value );

    // Storing a basic complex operation, multiple additions
    let sum_value = 1 + 2 + 3;
    println!( "The sum of 1 + 2 + 3 is {}", sum_value );

    // Complex logical operation with differing operation priorities
    let loyalty_member = false;
    let purchased5_coffees = true;
    let have_coupon = true;
    let coupon_not_expired = true;
    let free_coffee = loyalty_member && purchased5_coffees || have_coupon && coupon_not_expired;
    println!( "Customer gets a free coffee {}", free_coffee );

    // The code above is basically the same as the following, except the extra variables
    // Sometimes breaking apart complex statements can help to self document
    let loyalty_achieved = loyalty_member && purchased5_coffees;
    let have_valid_coupon = have_coupon && coupon_not_expired;
    let free_coffee = loyalty_achieved && have_valid_coupon;
    println!( "Customer gets a free coffee {}", free_coffee );

    // This code shows how a value will be computed incorrectly 
    // with default operator precedence
    let item_price = 99.99;
    let item_shipping = 9.99;
    let purchase_quantity = 5;
    let total_cost = item_price + item_shipping * purchase_quantity as f64;
    println!( "Total item cost is {:.2}", total_cost );

    // We can correct this with parenthesis which force operations to complete first
    let total_cost = (item_price + item_shipping) * purchase_quantity as f64;
    println!( "Total item cost is {:.2}", total_cost );

    // Parenthesis can be used to clarify precedence without 
    // having to know what the actual operator precedence is.
    let free_coffee = ((loyalty_member == true) && (purchased5_coffees == true)) || ((have_coupon == true) && (coupon_not_expired == true));

    // Some common errors, I can't really write these as a structured 
    // language program, so these are more pseudo code examples.
    // Do you want to assign 5 to a and b, or do you want to check if b is equal to 5?
    // Rust also does not allow a = b = 5, so this is less of a concern.
    let b = 5;
    let a = b;
    // vs
    // Rust throws a warning for these parenthesis, which I don't like as they clarify the statement.
    let aBool = (b == 5);

    // Spot the incorrect operator.
    // Rust's compiler catches this.
    free_coffee = ((loyalty_member = true) && (purchased5_coffees == true)) || ((have_coupon == true) && (coupon_not_expired == true));

    // This will not be isWeekend is true if day is either Saturday or Sunday
    let SATURDAY = 6;
    let SUNDAY = 7;
    let day = 6;
    // Rust will not compile with this, but not for the reason you might think.
    let is_weekend = (day == SATURDAY || SUNDAY);

    // This is what you should do
    let is_weekend = (day == SATURDAY) || (day == SUNDAY);
}
